#96 Create a Client operation to use TenantOptions when getting the tenant (JSDK-199)

maranai opened the pull request.

jarias commented on extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/tenant/TenantIT.groovy:
```diff
@@ -25,8 +25,11 @@ import com.stormpath.sdk.group.Group
 import com.stormpath.sdk.group.Groups
 import com.stormpath.sdk.provider.*
```
Don't know for the JSDK but for AM we have a code style policy of no * imports, not a big deal but we should keep our styles consistent.

maranai replied:
@jarias I've removed the * import. Please give it a look.

jarias commented on extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/tenant/TenantIT.groovy:
```diff
@@ -96,6 +99,39 @@ class TenantIT extends ClientIT {
+    /**
+     * @since 1.0.RC4.3-SNAPSHOT
+     */
+    @Test(enabled = false) //ignoring because of sporadic Travis failures
```
Has there been any research on why is this failing in Travis?

lhazlewood replied:
Not that I'm aware of

maranai replied:
@jarias and @lhazlewood initially, some tests failed because of performance problems. Many of them were fixed 2 or 3 months ago, but we still have some tests that fail sporadically.

maranai replied:
ITs were re-enabled as per our discussion with Les.

jarias commented:
Looks good

maranai commented:
@lhazlewood this PR was approved. If you have a moment, I would like you to take a look at it before I merge it. There's one test that fails for 1.6
Thanks!

maranai commented:
@mrioan this request has been approved, if you have a time, I would like you to take a look at it before I merge it.
Thanks!

mrioan commented:
@maranai sure. I'll take a look at it today

mrioan commented on api/src/main/java/com/stormpath/sdk/client/Client.java:
```diff
@@ -87,6 +88,15 @@
+     * @param tenantOptions The {@link TenantOptions} to use to customize the retrieved Tenant resource.
+     * @return the {@link Tenant} resource customized by the specified {@link TenantOptions}
+     *
+     * @since 1.0.RC4.3-SNAPSHOT
```
Current version is 1.0.RC4.6. BTW, versions must never have "-SNAPSHOT". SNAPSHOT is only used in `pom.xml` to denote that the code is still not stable.

mrioan commented on api/src/main/java/com/stormpath/sdk/ds/DataStore.java:
```diff
@@ -54,6 +55,23 @@
+     * @param <T>   type parameter indicating the returned value is a {@link Resource} instance.
+     * @return an instance of the specified class based on the data returned from the specified {@code href} URL and the specified {@code criteria}.
+     *
+     * @since 1.0.RC4.3-SNAPSHOT
```
Idem

mrioan replied:
Javadoc text should look similar to the one at `<T extends Resource, O extends Options> T getResource(String href, Class<T> clazz, O options);`. Please take a look at it in `get_resource_expanded` branch.

mrioan replied:
Why do you need this in the `DataStore`? This will be exposed to developers.

mrioan commented on api/src/main/java/com/stormpath/sdk/ds/DataStore.java:
```diff
@@ -54,6 +55,23 @@
+     *
+     * @since 1.0.RC4.3-SNAPSHOT
+     */
+    <T extends Resource> T getResource(String href, Class<T> clazz, Criteria criteria);
```
<T extends Resource, C extends Criteria> T getResource(String href, Class<T> clazz, C criteria);

mrioan commented on api/src/main/java/com/stormpath/sdk/tenant/TenantCriteria.java:
```diff
@@ -0,0 +1,34 @@
+ *  .withDirectories()
+ * </pre>
+ *
+ * @since 1.0.RC4.3-SNAPSHOT
```
idem

mrioan commented on extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/tenant/TenantIT.groovy:
```diff
@@ -96,6 +103,39 @@ class TenantIT extends ClientIT {
+
+        TenantOptions options = Tenants.options().withDirectories().withApplications()
+        def tenant = client.getCurrentTenant(options)
+        def apps = tenant.applications.size
```
This operation will cause the backend to be contacted again if the application is not expanded. This will give you the false idea that the expansion worked. So, you need to take a look at the internal properties of the tenant at this point rather than consulting it via SDK provided methods. Something like this:
```
Map properties = getValue(AbstractResource, tenant, "properties")
```
where

```
private Object getValue(Class clazz, Object object, String fieldName) {
    Field field = clazz.getDeclaredField(fieldName)
    field.setAccessible(true)
    return field.get(object)
}

```

mrioan commented on impl/src/main/java/com/stormpath/sdk/impl/tenant/DefaultTenantCriteria.java:
```diff
@@ -0,0 +1,72 @@
+import com.stormpath.sdk.tenant.TenantCriteria;
+import com.stormpath.sdk.tenant.TenantOptions;
+
+public class DefaultTenantCriteria extends DefaultCriteria<TenantCriteria, TenantOptions> implements TenantCriteria {
```
Please add version comment here

mrioan replied:
I think we need more tests to validate these new methods

maranai commented:
@mrioan changes have been applied and are ready for a new review.

mrioan commented:
@maranai I am realizing that `TenantOptions` is missing `accounts` and `groups`. Those collections are available in the `Tenant` endpoint in the backend. Therefore, we should also support them in `TenantOptions`. Can you please add them?

maranai commented:
@mrioan I've updated the PR and added some tests and changes that are required after new Tenant Options were created. Please take a look at this PR. Thanks!

mrioan commented:
@maranai back to you. I improved one of the tests to use the new `withGroups` option. If you are OK with my change then you can merge as I approve this PR. Thanks!

maranai commented:
I reviewed the test changes and I'm OK with them. I'll proceed with the merge.

mrioan commented:
Solves #96
